api/src/main/java/io/opentelemetry/common/ImmutableKeyValuePairs.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.common;
+
+import static io.opentelemetry.internal.Utils.checkArgument;
+import static io.opentelemetry.internal.Utils.checkNotNull;
+
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * An immutable set of key-value pairs. Keys are only {@link String} typed. Can be iterated over
+ * using the {@link #forEach(KeyValueConsumer)} method.
+ *
+ * @param <V> The type of the values contained in this.
+ * @see Labels
+ * @see Attributes
+ */
+@Immutable
+public interface ImmutableKeyValuePairs<V> {
```

Comment:
This can be package protected, no need to expose it public. Also why do you not make this an abstract class with all the utils inside it. It will remove some duplicate code between Labels and Attributes.

Reply:
I'll give it a try. Originally, I was hoping this could be a single-method-interface, and hence implementable as a lambda, but that won't be possible once we add `isEmpty()` and `size()`.

Reply:
Also, if we make this interface package-protected, we'll need to move the consumer interface up out of it. Not a big deal, but it is a consequence.

Reply:
That can be public correct.

Reply:
ok, update and pushed. PTaL

Reply:
i'm not super happy with the change to an abstract class, but we can address in the future.